README
NAME
    MooseX::CurriedDelegation - Curry your delegations with methods

VERSION
    This document describes version 0.002 of MooseX::CurriedDelegation -
    released May 19, 2014 as part of MooseX-CurriedDelegation.

SYNOPSIS
        use Moose;
        use MooseX::CurriedDelegation;

        has one => (is => 'ro', isa => 'Str', default => 'default');

        has foo => (

            is      => 'rw',
            isa     => 'TestClass::Delagatee', # has method curried()
            default => sub { TestClass::Delagatee->new },

            handles => {

                # method-curry
                #   Note the hashref, not arrayref, we employ
                #   first arg is the remote method to delegate to
                #   second is an arrayref comprising:
                #       coderef to call as a method on the instance, followed by
                #       "static" curry args
                #
                # so, essentially:
                #   $self->foo->remote_method($self->$coderef(), @remaining_args);
                #
                # foo_del_one => {
                #   remote_method => [ sub { ... }, qw{ static args } ],
                # },

                foo_del_one => { curried => [ sub { shift->one }, qw{ more curry args } ] },

                # curry_to_self() always returns: sub { shift }
                foo_del_two => { other_method => [ curry_to_self ] },
            },
        );

DESCRIPTION
    Method delegation is awfully handy -- but sometimes it'd be awfully
    handier if it was a touch more dynamic. This is an attribute trait that
    provides for a delegated method to be curried.

USAGE
    Using this package will cause the relevant attribute trait to be applied
    without requiring further intervention. We use the standard

        handles => { local_method => delegate_info, ... }

    delegation methodology, however our currying is invoked when
    delegate_info is a hashref. We expect delegate info to look like:

        { remote_method => [ $coderef, @more_curry_args ] }

    Only $coderef is ever executed by the delegation. This means that it is
    safe to have any number of additional coderefs in @more_curry_args: they
    will be passed through to remote_method without additional manipulation.

ADDITIONAL SUGAR
    In addition, we export a number of helper functions.

  curry_to_self()
    This function always returns a coderef like "sub { shift }". That is,
    this:

        local => { remote => [ curry_to_self ] }

    is equivalent to:

        local => { remote => [ sub { shift } ] }

TRAIT ALIASES
  CurriedDelegation
    Resolves out to the full name of our attribute trait; you can use it as:

        has foo => (traits => [CurriedDelegation], ...)

SEE ALSO
    Please see those modules/websites for more information related to this
    module.

    *   Moose

    *   Moose::Meta::Method::Delegation

SOURCE
    The development version is on github at
    <http://https://github.com/RsrchBoy/moosex-currieddelegation> and may be
    cloned from
    <git://https://github.com/RsrchBoy/moosex-currieddelegation.git>

BUGS
    Please report any bugs or feature requests on the bugtracker website
    https://github.com/RsrchBoy/moosex-currieddelegation/issues

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Chris Weyl <cweyl@alumni.drew.edu>

  I'm a material boy in a material world
    Please note I do not expect to be gittip'ed or flattr'ed for this work,
    rather it is simply a very pleasant surprise. I largely create and
    release works like this because I need them or I find it enjoyable;
    however, don't let that stop you if you feel like it ;)

    Flattr this
    <https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithu
    b.com%2FRsrchBoy%2Fmoosex-currieddelegation&title=RsrchBoy's%20CPAN%20Mo
    oseX-CurriedDelegation&tags=%22RsrchBoy's%20MooseX-CurriedDelegation%20i
    n%20the%20CPAN%22>, gittip me <https://www.gittip.com/RsrchBoy/>, or
    indulge my Amazon Wishlist <http://bit.ly/rsrchboys-wishlist>... If you
    so desire.

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2012 by Chris Weyl.

    This is free software, licensed under:

      The GNU Lesser General Public License, Version 2.1, February 1999



Changes
Revision history for MooseX-CurriedDelegation

{{$NEXT}}

0.003     2016-11-09 14:52:12 CST-0600
	* Fix our trait alias declaration (sigh)

0.002     2014-05-19 19:56:21 PDT-0700
    * Pod fixes, general readability fixes.

0.001     2012-01-25 17:29:24 America/Los_Angeles


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.017.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Curry your delegations with methods",
  "AUTHOR" => "Chris Weyl <cweyl\@alumni.drew.edu>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooseX-CurriedDelegation",
  "EXE_FILES" => [],
  "LICENSE" => "lgpl",
  "NAME" => "MooseX::CurriedDelegation",
  "PREREQ_PM" => {
    "Moose" => 0,
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "MooseX::Util" => 0,
    "aliased" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Moose" => 0,
    "Test::Moose::More" => 0,
    "Test::More" => "0.94"
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moose" => 0,
  "Moose::Exporter" => 0,
  "Moose::Role" => 0,
  "MooseX::Util" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Moose" => 0,
  "Test::Moose::More" => 0,
  "Test::More" => "0.94",
  "aliased" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/MooseX/CurriedDelegation.pm
package MooseX::CurriedDelegation;

# ABSTRACT: Curry your delegations with methods

use strict;
use warnings;

use Moose ( );
use Moose::Exporter;

my $trait = 'MooseX::CurriedDelegation::Trait::Attribute';

Moose::Exporter->setup_import_methods(
    trait_aliases => [ [ $trait => 'CurriedDelegation' ] ],

    as_is => [ qw{ curry_to_self } ],
    class_metaroles => { attribute         => [ $trait ] },
    role_metaroles  => { applied_attribute => [ $trait ] },
);

sub curry_to_self() { sub { shift } }

!!42;

__END__

=head1 SYNOPSIS

    use Moose;
    use MooseX::CurriedDelegation;

    has one => (is => 'ro', isa => 'Str', default => 'default');

    has foo => (

        is      => 'rw',
        isa     => 'TestClass::Delagatee', # has method curried()
        default => sub { TestClass::Delagatee->new },

        handles => {

            # method-curry
            #   Note the hashref, not arrayref, we employ
            #   first arg is the remote method to delegate to
            #   second is an arrayref comprising:
            #       coderef to call as a method on the instance, followed by
            #       "static" curry args
            #
            # so, essentially:
            #   $self->foo->remote_method($self->$coderef(), @remaining_args);
            #
            # foo_del_one => {
            #   remote_method => [ sub { ... }, qw{ static args } ],
            # },

            foo_del_one => { curried => [ sub { shift->one }, qw{ more curry args } ] },

            # curry_to_self() always returns: sub { shift }
            foo_del_two => { other_method => [ curry_to_self ] },
        },
    );


=head1 DESCRIPTION

Method delegation is awfully handy -- but sometimes it'd be awfully handier if
it was a touch more dynamic.  This is an attribute trait that provides for a
delegated method to be curried.

=head1 USAGE

Using this package will cause the relevant attribute trait to be applied
without requiring further intervention.  We use the standard

    handles => { local_method => delegate_info, ... }

delegation methodology, however our currying is invoked when delegate_info is
a hashref.  We expect delegate info to look like:

    { remote_method => [ $coderef, @more_curry_args ] }

Only $coderef is ever executed by the delegation.  This means that it is safe
to have any number of additional coderefs in @more_curry_args: they will be
passed through to remote_method without additional manipulation.

=head1 ADDITIONAL SUGAR

In addition, we export a number of helper functions.

=head2 curry_to_self()

This function always returns a coderef like "sub { shift }".  That is, this:

    local => { remote => [ curry_to_self ] }

is equivalent to:

    local => { remote => [ sub { shift } ] }

is equivalent to:

    $self->attribute_accessor->remote($self)

=head1 TRAIT ALIASES

=head2 CurriedDelegation

Resolves out to the full name of our attribute trait; you can use it as:

    has foo => (traits => [CurriedDelegation], ...)

=head1 SEE ALSO

L<Moose>
L<Moose::Meta::Method::Delegation>

=cut


lib/MooseX/CurriedDelegation/Trait/Attribute.pm
#
# This file is part of MooseX-CurriedDelegation
#
# This software is Copyright (c) 2012 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
package MooseX::CurriedDelegation::Trait::Attribute;
BEGIN {
  $MooseX::CurriedDelegation::Trait::Attribute::AUTHORITY = 'cpan:RSRCHBOY';
}
$MooseX::CurriedDelegation::Trait::Attribute::VERSION = '0.002';
# ABSTRACT: The great new MooseX::CurriedDelegation!

use Moose::Role;
use namespace::autoclean;
use MooseX::Util 'with_traits';

use aliased 'MooseX::CurriedDelegation::Trait::Method::Delegation' => 'OurMethodTrait';



sub method_curried_delegation_metaclass {
    my $self = shift @_;

    return with_traits($self->delegation_metaclass => OurMethodTrait);
}

around _make_delegation_method => sub {
    my ($orig, $self) = (shift, shift);
    my ($handle_name, $delegation_hash) = @_;

    ### $delegation_hash
    return $self->$orig(@_)
        unless 'HASH' eq ref $delegation_hash;

    confess "We should never have more than one key when setting up delegation: $handle_name"
        if keys %$delegation_hash > 1;

    my ($method_to_call) = keys %$delegation_hash;
    my @curry_args       = @{ $delegation_hash->{$method_to_call} };
    my $curry_coderef    = shift @curry_args;

    ### $method_to_call
    ### @curry_args

    return $self->method_curried_delegation_metaclass->new(
        name               => $handle_name,
        package_name       => $self->associated_class->name,
        attribute          => $self,
        delegate_to_method => $method_to_call,
        curry_coderef      => $curry_coderef,
        curried_arguments  => \@curry_args,
    );
};

!!42;

__END__

=pod

=encoding UTF-8

=for :stopwords Chris Weyl

=for :stopwords Wishlist flattr flattr'ed gittip gittip'ed

=head1 NAME

MooseX::CurriedDelegation::Trait::Attribute - The great new MooseX::CurriedDelegation!

=head1 VERSION

This document describes version 0.002 of MooseX::CurriedDelegation::Trait::Attribute - released May 19, 2014 as part of MooseX-CurriedDelegation.

=head1 DESCRIPTION

This is just a trait applied to the delegation method metaclass (generally
L<Moose::Meta::Method::Delegation>).  No user-serviceable parts here.

=head1 METHODS

=head2 method_curried_delegation_metaclass

Returns a class composed of the applied class' delegation metaclass and our
delegation method trait.

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<MooseX::CurriedDelegation|MooseX::CurriedDelegation>

=item *

L<MooseX::CurriedDelegation>

=back

=head1 SOURCE

The development version is on github at L<http://https://github.com/RsrchBoy/moosex-currieddelegation>
and may be cloned from L<git://https://github.com/RsrchBoy/moosex-currieddelegation.git>

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/RsrchBoy/moosex-currieddelegation/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head2 I'm a material boy in a material world

=begin html

<a href="https://www.gittip.com/RsrchBoy/"><img src="https://raw.githubusercontent.com/gittip/www.gittip.com/master/www/assets/%25version/logo.png" /></a>
<a href="http://bit.ly/rsrchboys-wishlist"><img src="http://wps.io/wp-content/uploads/2014/05/amazon_wishlist.resized.png" /></a>
<a href="https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-currieddelegation&title=RsrchBoy's%20CPAN%20MooseX-CurriedDelegation&tags=%22RsrchBoy's%20MooseX-CurriedDelegation%20in%20the%20CPAN%22"><img src="http://api.flattr.com/button/flattr-badge-large.png" /></a>

=end html

Please note B<I do not expect to be gittip'ed or flattr'ed for this work>,
rather B<it is simply a very pleasant surprise>. I largely create and release
works like this because I need them or I find it enjoyable; however, don't let
that stop you if you feel like it ;)

L<Flattr this|https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-currieddelegation&title=RsrchBoy's%20CPAN%20MooseX-CurriedDelegation&tags=%22RsrchBoy's%20MooseX-CurriedDelegation%20in%20the%20CPAN%22>,
L<gittip me|https://www.gittip.com/RsrchBoy/>, or indulge my
L<Amazon Wishlist|http://bit.ly/rsrchboys-wishlist>...  If you so desire.

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Chris Weyl.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999

=cut


lib/MooseX/CurriedDelegation/Trait/Method/Delegation.pm
#
# This file is part of MooseX-CurriedDelegation
#
# This software is Copyright (c) 2012 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
package MooseX::CurriedDelegation::Trait::Method::Delegation;
BEGIN {
  $MooseX::CurriedDelegation::Trait::Method::Delegation::AUTHORITY = 'cpan:RSRCHBOY';
}
$MooseX::CurriedDelegation::Trait::Method::Delegation::VERSION = '0.002';
# ABSTRACT: A trait for curried delegation methods

use Moose::Role;
use namespace::autoclean;

# debugging...
#use Smart::Comments;

has curry_coderef => (is => 'ro', isa => 'Coderef', required => 1);

# _initialize_body() is largely lifted right from
# Moose::Meta::Method::Delegation

sub _initialize_body {
    my $self = shift;

    my $method_to_call = $self->delegate_to_method;
    # XXX
    #return $self->{body} = $method_to_call
    #    if ref $method_to_call;

    my $accessor      = $self->_get_delegate_accessor;
    my $handle_name   = $self->name;

    return $self->{body} = sub {

        my $instance      = shift;
        my $proxy         = $instance->$accessor();
        my $curry_coderef = $self->curry_coderef;

        ### $curry_coderef

        my $error
            = !defined $proxy                 ? ' is not defined'
            : ref($proxy) && !blessed($proxy) ? qq{ is not an object (got '$proxy')}
            : undef;

        if ($error) {
            $self->throw_error(
                "Cannot delegate $handle_name to $method_to_call because "
                    . "the value of "
                    . $self->associated_attribute->name
                    . $error,
                method_name => $method_to_call,
                object      => $instance
            );
        }
        unshift @_, @{ $self->curried_arguments };
        unshift @_, $instance->$curry_coderef();
        $proxy->$method_to_call(@_);
    };
}

!!42;

__END__

=pod

=encoding UTF-8

=for :stopwords Chris Weyl

=for :stopwords Wishlist flattr flattr'ed gittip gittip'ed

=head1 NAME

MooseX::CurriedDelegation::Trait::Method::Delegation - A trait for curried delegation methods

=head1 VERSION

This document describes version 0.002 of MooseX::CurriedDelegation::Trait::Method::Delegation - released May 19, 2014 as part of MooseX-CurriedDelegation.

=head1 DESCRIPTION

This is just a trait applied to the delegation method metaclass (generally
L<Moose::Meta::Method::Delegation>).  No user-serviceable parts here.

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<MooseX::CurriedDelegation|MooseX::CurriedDelegation>

=item *

L<MooseX::CurriedDelegation>

=back

=head1 SOURCE

The development version is on github at L<http://https://github.com/RsrchBoy/moosex-currieddelegation>
and may be cloned from L<git://https://github.com/RsrchBoy/moosex-currieddelegation.git>

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/RsrchBoy/moosex-currieddelegation/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head2 I'm a material boy in a material world

=begin html

<a href="https://www.gittip.com/RsrchBoy/"><img src="https://raw.githubusercontent.com/gittip/www.gittip.com/master/www/assets/%25version/logo.png" /></a>
<a href="http://bit.ly/rsrchboys-wishlist"><img src="http://wps.io/wp-content/uploads/2014/05/amazon_wishlist.resized.png" /></a>
<a href="https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-currieddelegation&title=RsrchBoy's%20CPAN%20MooseX-CurriedDelegation&tags=%22RsrchBoy's%20MooseX-CurriedDelegation%20in%20the%20CPAN%22"><img src="http://api.flattr.com/button/flattr-badge-large.png" /></a>

=end html

Please note B<I do not expect to be gittip'ed or flattr'ed for this work>,
rather B<it is simply a very pleasant surprise>. I largely create and release
works like this because I need them or I find it enjoyable; however, don't let
that stop you if you feel like it ;)

L<Flattr this|https://flattr.com/submit/auto?user_id=RsrchBoy&url=https%3A%2F%2Fgithub.com%2FRsrchBoy%2Fmoosex-currieddelegation&title=RsrchBoy's%20CPAN%20MooseX-CurriedDelegation&tags=%22RsrchBoy's%20MooseX-CurriedDelegation%20in%20the%20CPAN%22>,
L<gittip me|https://www.gittip.com/RsrchBoy/>, or indulge my
L<Amazon Wishlist|http://bit.ly/rsrchboys-wishlist>...  If you so desire.

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Chris Weyl.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999

=cut
